staticjinja/staticjinja.py

```diff
@@ -93,39 +117,40 @@ def get_template(self, template_name):
         """
         return self._env.get_template(template_name)
 
-    def _get_context_generator(self, template_name):
-        """Get a context generator for a template.
-
-        Raises a :exc:`ValueError` if no matching context generator can be
-        found.
-
-        :param template_name: the name of the template
-        """
-        for regex, context_generator in self.contexts:
-            if re.match(regex, template_name):
-                return context_generator
-        raise ValueError("no matching context generator")
-
     def get_context(self, template):
         """Get the context for a template.
 
         By default, this function will return an empty context.
 
-        If a matching context_generator can be found, it will be passed the
-        template and executed. The context generator should return a dictionary
-        representing the context.
+        If a matching context_generator can be found and is a function
+        it will be executed after being passed the template if it expects an
+        argument. If it is not a function then it will be assumed to act like
+        a dictionary object.
+
+        If several context_generator can be found, their resulting dictionary
+        will be merged if mergecontexts is True or only the first one will be
+        used otherwise.
 
         :param template: the template to get the context for
         """
-        try:
-            context_generator = self._get_context_generator(template.name)
-        except ValueError:
-            return {}
-        else:
-            try:
-                return context_generator(template)
-            except TypeError:
-                return context_generator()
+
+        context = {}
+        for regex, context_generator in self.contexts:
+            if re.match(regex, template.name):
+                if inspect.isfunction(context_generator):
+                    if _has_argument(context_generator):
+                        context.update(context_generator(template))
+                    else:
+                        context.update(context_generator())
```

Comment:
Hm, yeah you have me convinced that catching every `TypeError` is bad. However, I wonder if it a better solution to this would be to always pass the template to the context generator. I'm not sure why I designed it to be optional. I'm guessing because I disliked the idea of passing a template to a function which wouldn't need it, but I can't remember. I'm not sure such a design has any precedent in Python. Do you agree here?

Obviously, always passing templates to context generators would be backwards incompatible, but this change makes sense as an intermediate step to help improve type safety for existing users. We should warn people in the case that they have context generators that don't accept templates though and then adjust the docs.